feat: short answer element for question editor

joyce-shi opened the pull request:
## Notion ticket link
<!-- Please replace with your ticket's URL -->
[Add short answer component to Question Editor](https://www.notion.so/uwblueprintexecs/Add-short-answer-component-to-Question-Editor-be5f17d52b9148f0918cf276c71359ad?pvs=4)


<!-- Give a quick summary of the implementation details, provide design justifications if necessary -->
## Implementation description
* Adding short answer component to Question Editor

<!-- Draw attention to the substantial parts of your PR or anything you'd like a second opinion on -->
## What should reviewers focus on?
* feature works as expected
* code cleanliness
* cosmetic issues

joyce-shi committed:
cherry-picked

joyce-shi committed:
cosmetic fixes

joyce-shi committed:
add edit icon

joyce-shi committed:
edit funcionality

joyce-shi committed:
making error optional in helper function

joyce-shi commented on frontend/src/components/question-creation/question-elements/modals/short-answer/ShortAnswerModal.tsx:
```diff
@@ -0,0 +1,86 @@
+          </ModalFooter>
+        </ModalContent>
+      </Modal>
+    </>
```
I'll probably refactor this modal into a general `ResponseModal` element in the future so that we can use it for all the response types in the question editor.

joyce-shi committed:
updating naming

joyce-shi commented on frontend/src/components/question-creation/question-elements/modals/short-answer/ShortAnswerModal.tsx:
```diff
@@ -0,0 +1,86 @@
+    } else {
+      setError(true);
+    }
+  };
```
i hate this function, so please lmk if you have ideas on how to class it up bit :P

joyce-shi commented on frontend/src/components/pages/admin/QuestionPage.tsx:
```diff
@@ -10,20 +10,31 @@ import QuestionEditor from "../../question-creation/QuestionEditor";
+          setQuestionElements,
+          showAddShortAnswerModal,
+          setShowAddShortAnswerModal,
+        }}
```
i'm not sure if i love having this in the editor context... but I'm not sure if there's a better place to put this

joyce-shi commented on frontend/src/components/pages/admin/QuestionPage.tsx:
```diff
@@ -10,20 +10,31 @@ import QuestionEditor from "../../question-creation/QuestionEditor";
           <QuestionSidebar backPage={HOME_PAGE} />
           <QuestionEditor />
         </Flex>
+        <AddShortAnswerModal />
```
same with having this modal here... but idk where else to put it

joyce-shi commented on frontend/src/components/question-creation/question-elements/ShortAnswerElement.tsx:
```diff
@@ -0,0 +1,51 @@
+  return (
+    <Flex width="100%" paddingLeft="6">
+      <Input
+        value={data}
```
differs from the Figma designs, but design team has asked to have the answer value shown here

joyce-shi commented on frontend/src/components/question-creation/question-elements/ShortAnswerElement.tsx:
```diff
@@ -0,0 +1,51 @@
+        borderRadius="0"
+        borderColor="grey.300"
+        focusBorderColor="grey.300"
+        width="34%"
```
fixed width for all short answer elements, as per design

joyce-shi committed:
spacing

joyce-shi commented on frontend/src/components/question-creation/question-elements/modals/short-answer/ShortAnswerModal.tsx:
```diff
@@ -0,0 +1,86 @@
+        <ModalOverlay />
+        <ModalContent minWidth="42vw">
+          <ModalHeader>
+            <Text textStyle="subtitle2">Create multi-select question</Text>
```
```suggestion
            <Text textStyle="subtitle2">Create short answer question</Text>
```

joyce-shi committed:
Update frontend/src/components/question-creation/question-elements/modals/short-answer/ShortAnswerModal.tsx

joyce-shi committed:
fixing reset data bug

joyce-shi commented on frontend/src/components/question-creation/question-elements/modals/short-answer/ShortAnswerModal.tsx:
```diff
@@ -0,0 +1,93 @@
+          </ModalBody>
+          <Divider color="grey.200" style={{ marginTop: "1.5em" }} />
+          <ModalFooter>
+            <Button variant="secondary" onClick={onClose} minWidth="10%" mr={2}>
```
```suggestion
            <Button variant="secondary" onClick={handleClose} minWidth="10%" mr={2}>
```

joyce-shi committed:
Update frontend/src/components/question-creation/question-elements/modals/short-answer/ShortAnswerModal.tsx

carissa-tang commented on frontend/src/components/question-creation/question-elements/modals/short-answer/ShortAnswerModal.tsx:
```diff
@@ -0,0 +1,93 @@
+      onConfirm(answer);
+      setAnswer("");
+      setError(false);
+      onClose();
```
```suggestion
      handleClose();
```
need to define `handleClose` before `handleConfirm`

carissa-tang commented on frontend/src/components/question-creation/question-elements/modals/short-answer/ShortAnswerModal.tsx:
```diff
@@ -0,0 +1,93 @@
+  };
+
+  return (
+    <>
```
```suggestion
```
can remove empty tags

carissa-tang commented on frontend/src/components/question-creation/question-elements/ShortAnswerElement.tsx:
```diff
@@ -0,0 +1,51 @@
+
+interface ShortAnswerElementProps {
+  id: string;
+  data: string;
```
is there a reason we store it as a string instead of number?

carissa-tang replied:
also inputs such as `01` are allowed, not sure how we want to handle that

joyce-shi replied:
updated to store as number! should resolve the `01` issue.

carissa-tang reviewed (commented):
left some comments!!

i noticed this issue with the data not showing in the modal after editing the input 1+ times. could you look into it?
edit: i think its because `<AddShortAnswerModal/>` gets rendered again in `QuestionPage.tsx` and there's no `data` passed so it gets set to undefined

https://user-images.githubusercontent.com/69697180/222637437-7f34a210-e353-406d-b6d2-472d78f6b99c.mov

joyce-shi committed:
pr comments

joyce-shi committed:
bug fix

joyce-shi commented:
> left some comments!!
> 
> i noticed this issue with the data not showing in the modal after editing the input 1+ times. could you look into it? edit: i think its because `<AddShortAnswerModal/>` gets rendered again in `QuestionPage.tsx` and there's no `data` passed so it gets set to undefined

should be resolved now! let me know if you find any further issues w/ it (: 
it's bc i was setting the initial state of the comment = to `data`, but initial state of the comment only gets set when the component initially renders... so props value & state values were out of sync, oops
- added `useEffect` to sync props and state
- updated `setAnswer(undefined)` on close -> `setAnswer(data)` to reset the modal to the initial value (instead of a null value)

carissa-tang commented on frontend/src/components/question-creation/question-elements/modals/short-answer/ShortAnswerModal.tsx:
```diff
@@ -0,0 +1,104 @@
+            </FormErrorMessage>
+          </FormControl>
+        </ModalBody>
+        <Divider color="grey.200" style={{ marginTop: "1.5em" }} />
```
```suggestion
        <Divider color="grey.200" mt="1.5em" />
```

carissa-tang commented on frontend/src/components/question-creation/question-elements/modals/short-answer/ShortAnswerModal.tsx:
```diff
@@ -0,0 +1,104 @@
+            <FormLabel color="grey.300" style={{ fontSize: "18px" }}>
+              Enter correct answer
+            </FormLabel>
+            <Input
```
seems like now we can't take in decimals

joyce-shi replied:
hmmm but do students really need to learn about decimals? 🤔

joyce-shi replied:
just kidding ill take a look

carissa-tang reviewed (commented):
one more comment ! sorryy haha

joyce-shi committed:
Update frontend/src/components/question-creation/question-elements/modals/short-answer/ShortAnswerModal.tsx

Co-authored-by: Carissa Tang <69697180+carissa-tang@users.noreply.github.com>

joyce-shi committed:
fix float issue
